api/src/task/notifications/closing_date_notification.py

```diff
@@ -0,0 +1,185 @@
+                    "{CONTACT_INFO}"
+                )
+                notification[user_id] = message
+
+            elif len(saved_items) > 1:
+                # Multiple opportunities closing
+                message = (
+                    "Applications for the following funding opportunities are due in two weeks:\n\n"
+                )
+
+                for opportunity in saved_items:
+                    close_date_stmt = select(OpportunitySummary.close_date).where(
+                        OpportunitySummary.opportunity_id == opportunity.opportunity_id
+                    )
+                    close_date = self.db_session.execute(close_date_stmt).scalar_one_or_none()
+                    if close_date:
+                        message += (
+                            f"[{opportunity.opportunity.opportunity_title}]\n"
+                            f"Application due date: {close_date.strftime('%B %d, %Y')}\n\n"
+                        )
+
+                message += (
+                    "Please carefully review the opportunity listings for all requirements and deadlines.\n\n"
+                    "Sign in to Simpler.Grants.gov to manage your bookmarked opportunities.\n\n"
+                    "If you have questions, please contact the Grants.gov Contact Center:\n"
+                    "{CONTACT_INFO}"
+                )
+
+                notification[user_id] = message
+
+        return EmailData(
+            to_addresses=to_address_list,
+            subject="Applications for your bookmarked funding opportunities are due soon",
+            content=notification,
+            notification_reason=NotificationReasons.CLOSING_DATE_REMINDER,
+        )
+
+    def create_user_opportunity_notification_log(self) -> None:
+        if self.collected_data:
+            for user_id, saved_items in self.collected_data.items():
+                for closing_opportunity in saved_items:
+                    if len(closing_opportunity) > 0:
+                        # Create notification log entry
+                        opp_notification_log = UserOpportunityNotificationLog(
+                            user_id=user_id,
+                            opportunity_id=closing_opportunity.opportunity_id,
+                        )
+                        self.db_session.add(opp_notification_log)
+
+                logger.info(
+                    "Successfully sent closing date reminder",
+                    extra={
+                        "user_id": user_id,
+                        "opportunity_ids": [opp.opportunity_id for opp in saved_items],
+                    },
+                )
+
+    def run_task(self) -> None:
+        """Override to define the task logic"""
+        data = self.notification_data()
+        if data:
+            self.send_notifications(data, self.pinpoint_client, self.app_id)
+            self.create_user_opportunity_notification_log()
```

Comment:
Make sure to put `with db_session.begin():` around whatever we're running so it commits at the end.